Source/dthread.cpp

```diff
@@ -4,131 +4,108 @@
  * Implementation of functions for updating game state from network commands.
  */
 
+#include <list>
+#include <atomic>
+
 #include "nthread.h"
 #include "storm/storm.h"
 #include "utils/thread.h"
 
 namespace devilution {
 
+struct DThreadPkt {
+	int pnum;
+	_cmd_id cmd;
+	std::unique_ptr<byte[]> data;
+	uint32_t len;
+
+	DThreadPkt(int pnum, _cmd_id(cmd), std::unique_ptr<byte[]> data, uint32_t len)
+		: pnum(pnum)
+		, cmd(cmd)
+		, data(std::move(data))
+		, len(len)
+	{
+	}
+};
+
 namespace {
 
 CCritSect sgMemCrit;
 SDL_threadID glpDThreadId;
-TMegaPkt *sgpInfoHead; /* may not be right struct */
-bool dthread_running;
+std::list<DThreadPkt> InfoList;
+std::atomic_bool dthread_running;
 event_emul *sghWorkToDoEvent;
 
 /* rdata */
 SDL_Thread *sghThread = nullptr;
 
 void DthreadHandler()
 {
-	const char *errorBuf;
-	TMegaPkt *pkt;
-	DWORD dwMilliseconds;
-
 	while (dthread_running) {
-		if (sgpInfoHead == nullptr && WaitForEvent(sghWorkToDoEvent) == -1) {
-			errorBuf = SDL_GetError();
-			app_fatal("dthread4:\n%s", errorBuf);
-		}
+		if (InfoList.empty() && WaitForEvent(sghWorkToDoEvent) == -1)
+			app_fatal("dthread4:\n%s", SDL_GetError());
 
 		sgMemCrit.Enter();
-		pkt = sgpInfoHead;
-		if (sgpInfoHead != nullptr)
-			sgpInfoHead = sgpInfoHead->pNext;
-		else
+		if (InfoList.empty()) {
 			ResetEvent(sghWorkToDoEvent);
+			sgMemCrit.Leave();
+			continue;
+		}
+		DThreadPkt pkt = std::move(InfoList.front());
+		InfoList.pop_front();
 		sgMemCrit.Leave();
 
-		if (pkt != nullptr) {
-			if (pkt->dwSpaceLeft != MAX_PLRS)
-				multi_send_zero_packet(pkt->dwSpaceLeft, static_cast<_cmd_id>(pkt->data[0]), &pkt->data[8], *(DWORD *)&pkt->data[4]);
+		if (pkt.pnum != MAX_PLRS)
+			multi_send_zero_packet(pkt.pnum, pkt.cmd, pkt.data.get(), pkt.len);
 
-			dwMilliseconds = 1000 * *(DWORD *)&pkt->data[4] / gdwDeltaBytesSec;
-			if (dwMilliseconds >= 1)
-				dwMilliseconds = 1;
+		DWORD dwMilliseconds = 1000 * pkt.len / gdwDeltaBytesSec;
+		if (dwMilliseconds >= 1)
+			dwMilliseconds = 1;
 
-			std::free(pkt);
-
-			if (dwMilliseconds != 0)
-				SDL_Delay(dwMilliseconds);
-		}
+		if (dwMilliseconds != 0)
+			SDL_Delay(dwMilliseconds);

```

Comment:
I'm have no idea what the rationale for this sleep is. Further, this also sleeps if it's silently processing a packet pertaining to a player that has left in the meantime.

As for switching to standard C++ stuff, I'm all for it. There does seem to be a bias toward SDL stuff in this project. (SDL_Mutex having to be managed manually as a pointer is particularly annoying.) Maybe there are portability issues. I haven't really poked in this direction.

Reply:
There's a really good chance 3DS would need to use `SDL_Delay()`.

Reply:
I have a suspicion that the delay is a leftover from when they were using modems. Nothing in the network API suggests that it is needed; stuff can be buffered without any issue.

I say we do away with it.

Reply:
It looks like it's related to `caps.bytessec` defined by `SNetGetProviderCaps()`.
https://github.com/diasurgical/devilutionX/blob/ba60907853ebc0804ab7143ba71ad5652d1f7af4/Source/nthread.cpp#L166

If you assume that `caps.bytessec` and `gdwDeltaBytesSec` are each in units of bytes per second, the equation to calculate dwMilliseconds seems reasonable. What's odd is that the limit is reduced by a factor of 4, then the delay is forced down to the lowest possible wait time. Perhaps Blizzard expected that `Sleep(1)` would typically pause for around 250 milliseconds or something.

Anyway, it looks to me like a rather feeble attempt to enable a provider to limit bandwidth. I agree that we should do away with it for now. If we actually feel we need to cap bandwidth for some reason in the future, it's probably best to start again from scratch.

Reply:
This is also the only place in the code that uses gdwDeltaBytesSec.

Reply:
Yielding in `while(dthread_running) { ... }` does have an advantage of letting the OS perform useful work when this has nothing to do. The previous implementation didn't handle that well though since it only used sleep after it had received a packet. Also the OS scheduler might background the thread while it's acquiring the lock on `sgMemCrit`?

Reply:
`WaitForEvent()` uses a condition variable to wait for a signal that indicates it has something to do. If it has nothing to do, it just waits on the condition variable. The delay was only being used to limit bandwidth when the thread has something to do.

Reply:
Good point, no need to do another yield in that case.